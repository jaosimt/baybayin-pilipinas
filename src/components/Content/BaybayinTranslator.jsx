import React from 'react';
import './BaybayinTranslator.scss';
import {isMobile} from "../../utils";

export default class BaybayinTranslator extends React.Component {
    ___bybyn = {
        a: 'ᜀ',
        e: 'ᜁ',
        i: 'ᜁ',
        o: 'ᜂ',
        u: 'ᜂ',
        
        b: 'ᜊ᜔',
        ba: 'ᜊ',
        be: 'ᜊᜒ',
        bi: 'ᜊᜒ',
        bo: 'ᜊᜓ',
        bu: 'ᜊᜓ',
        
        k: 'ᜃ᜔',
        ka: 'ᜃ',
        ke: 'ᜃᜒ',
        ki: 'ᜃᜒ',
        ko: 'ᜃᜓ',
        ku: 'ᜃᜓ',
        
        d: 'ᜇ᜔',
        da: 'ᜇ',
        de: 'ᜇᜒ',
        di: 'ᜇᜒ',
        do: 'ᜇᜓ',
        du: 'ᜇᜓ',
        
        g: 'ᜄ᜔',
        ga: 'ᜄ',
        ge: 'ᜄᜒ',
        gi: 'ᜄᜒ',
        go: 'ᜄᜓ',
        gu: 'ᜄᜓ',
        
        h: 'ᜑ᜔',
        ha: 'ᜑ',
        he: 'ᜑᜒ',
        hi: 'ᜑᜒ',
        ho: 'ᜑᜓ',
        hu: 'ᜑᜓ',
        
        l: 'ᜎ᜔',
        la: 'ᜎ',
        le: 'ᜎᜒ',
        li: 'ᜎᜒ',
        lo: 'ᜎᜓ',
        lu: 'ᜎᜓ',
        
        m: 'ᜋ᜔',
        ma: 'ᜋ',
        me: 'ᜋᜒ',
        mi: 'ᜋᜒ',
        mo: 'ᜋᜓ',
        mu: 'ᜋᜓ',
        
        n: 'ᜈ᜔',
        na: 'ᜈ',
        ne: 'ᜈᜒ',
        ni: 'ᜈᜒ',
        no: 'ᜈᜓ',
        nu: 'ᜈᜓ',
        
        ng: 'ᜅ᜔',
        nga: 'ᜅ',
        nge: 'ᜅᜒ',
        ngi: 'ᜅᜒ',
        ngo: 'ᜅᜓ',
        ngu: 'ᜅᜓ',
        
        p: 'ᜉ᜔',
        pa: 'ᜉ',
        pe: 'ᜉᜒ',
        pi: 'ᜉᜒ',
        po: 'ᜉᜓ',
        pu: 'ᜉᜓ',
        
        r: 'ᜍ᜔',
        ra: 'ᜍ',
        re: 'ᜍᜒ',
        ri: 'ᜍᜒ',
        ro: 'ᜍᜓ',
        ru: 'ᜍᜓ',
        
        s: 'ᜐ᜔',
        sa: 'ᜐ',
        se: 'ᜐᜒ',
        si: 'ᜐᜒ',
        so: 'ᜐᜓ',
        su: 'ᜐᜓ',
        
        t: 'ᜆ᜔',
        ta: 'ᜆ',
        te: 'ᜆᜒ',
        ti: 'ᜆᜒ',
        to: 'ᜆᜓ',
        tu: 'ᜆᜓ',
        
        w: 'ᜏ᜔',
        wa: 'ᜏ',
        we: 'ᜏᜒ',
        wi: 'ᜏᜒ',
        wo: 'ᜏᜓ',
        wu: 'ᜏᜓ',
        
        y: 'ᜌ᜔',
        ya: 'ᜌ',
        ye: 'ᜌᜒ',
        yi: 'ᜌᜒ',
        yo: 'ᜌᜓ',
        yu: 'ᜌᜓ',
        
        c: 'c',
        f: 'f',
        j: 'j',
        q: 'q',
        v: 'v',
        x: 'x',
        z: 'z'
    };
    
    ___ynbyby = {
        "ᜂ": "[o/u]",
        "ᜁ": "[i/e]",
        "ᜀ": "a",
        
        "ᜊᜓ": "b[o/u]",
        "ᜊᜒ": "b[i/e]",
        "ᜊ᜔": "b",
        "ᜊ": "ba",
        
        "ᜃᜓ": "k[o/u]",
        "ᜃᜒ": "k[i/e]",
        "ᜃ᜔": "k",
        "ᜃ": "ka",
        
        "ᜇᜓ": "d[o/u]",
        "ᜇᜒ": "d[i/e]",
        "ᜇ᜔": "d",
        "ᜇ": "da",
        
        "ᜄᜓ": "g[o/u]",
        "ᜄᜒ": "g[i/e]",
        "ᜄ᜔": "g",
        "ᜄ": "ga",
        
        "ᜑᜓ": "h[o/u]",
        "ᜑᜒ": "h[i/e]",
        "ᜑ᜔": "h",
        "ᜑ": "ha",
        
        "ᜎᜓ": "l[o/u]",
        "ᜎᜒ": "l[i/e]",
        "ᜎ᜔": "l",
        "ᜎ": "la",
        
        "ᜋᜓ": "m[o/u]",
        "ᜋᜒ": "m[i/e]",
        "ᜋ᜔": "m",
        "ᜋ": "ma",
        
        "ᜈᜓ": "n[o/u]",
        "ᜈᜒ": "n[i/e]",
        "ᜈ᜔": "n",
        "ᜈ": "na",
        
        "ᜅᜓ": "ng[o/u]",
        "ᜅᜒ": "ng[i/e]",
        "ᜅ᜔": "ng",
        "ᜅ": "nga",
        
        "ᜉᜓ": "p[o/u]",
        "ᜉᜒ": "p[i/e]",
        "ᜉ᜔": "p",
        "ᜉ": "pa",
        
        "ᜍᜓ": "r[o/u]",
        "ᜍᜒ": "r[i/e]",
        "ᜍ᜔": "r",
        "ᜍ": "ra",
        
        "ᜐᜓ": "s[o/u]",
        "ᜐᜒ": "s[i/e]",
        "ᜐ᜔": "s",
        "ᜐ": "sa",
        
        "ᜆᜓ": "t[o/u]",
        "ᜆᜒ": "t[i/e]",
        "ᜆ᜔": "t",
        "ᜆ": "ta",
        
        "ᜏᜓ": "w[o/u]",
        "ᜏᜒ": "w[i/e]",
        "ᜏ᜔": "w",
        "ᜏ": "wa",
        
        "ᜌᜓ": "y[o/u]",
        "ᜌᜒ": "y[i/e]",
        "ᜌ᜔": "y",
        "ᜌ": "ya"
    };
    
    previewMessage = `${isMobile() ? '↑↑↑' : '← ← ←'} type or paste something there<br/>and see the translation here. ↓↓↓`;
    
    state = {
        baybayin: '',
        translation: this.previewMessage,
        simplify: true
    };
    
    onTextAreaChange = (e) => {
        e.target.value = e.target.value.toLowerCase().replace(/(ᜊ᜔|ᜃ᜔|ᜇ᜔|ᜄ᜔|ᜑ᜔|ᜎ᜔|ᜋ᜔|ᜈ᜔|ᜅ᜔|ᜉ᜔|ᜍ᜔|ᜐ᜔|ᜆ᜔|ᜏ᜔|ᜌ᜔)([aieoug])/ig, (x, y, z) => {
            return this.___bybyn[this.___ynbyby[y] + z] || x
        }).replace(/ng([aieou])*/ig, (x) => {
            return this.___bybyn[x] || x;
        }).replace(/([bkdghlmnprstwy])([aieou])/ig, (x) => {
            return this.___bybyn[x] || x;
        }).replace(/[a-z]/ig, (x) => {
            return this.___bybyn[x] || x;
        });
        
        this.setState({
            baybayin: e.target.value,
            translation: this.translate(e.target.value) || this.previewMessage
        })
    };
    
    translate = (str, normalize) => {
        str = str.toString();
        if (str.trim() === '') return;
        
        for (let b in this.___ynbyby) {
            if (this.___ynbyby.hasOwnProperty(b)) str = str.replace(new RegExp(b, 'igm'), this.___ynbyby[b]);
        }
        
        return ((normalize === true || this.state.simplify.checked) ? str.replace(/\[e\/i\]/g, 'i').replace(/\[o\/u\]/g, 'o') : str).replace(/(\[\w\/\w\])/gm, "<span class='dimmed'>$1</span>").replace(/\n/gm, '<br/>');
    };
    
    componentDidMount() {
        if (this.node) this.node.focus();
    };
    
    render() {
        return <div className={'baybayin-translator'}>
            <textarea
                ref={node => this.node = node}
                defaultValue={this.state.baybayin}
                onKeyUp={this.onTextAreaChange}
            />
            
            <div className={'preview'} dangerouslySetInnerHTML={{__html: this.state.translation}}/>
            
            <div className={'info'}>
            </div>
        </div>;
    }
}