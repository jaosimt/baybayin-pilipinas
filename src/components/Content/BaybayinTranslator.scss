@import "../../common";
@import "./Baybayin-Modern-Club";

.baybayin-translator {
    position: relative;
    display: grid;
    grid-template-columns: 50% 50% auto;
    grid-template-areas:
        'text-area preview'
        'info info'; // chart';
    grid-column-gap: $grid-gap;
    height: 100%;
    
    textarea,
    .preview {
        background-color: white;
        -webkit-border-radius: 14px;
        -moz-border-radius: 14px;
        border-radius: 14px;
        min-height: 200px;
    }
    
    //.chart,
    textarea,
    .preview {
        padding: 14px;
    }
    
    textarea {
        grid-area: text-area;
        resize: none;
        font-size: 49px;
        font-family: Baybayin-Modern-Club;
    }
    
    .preview {
        grid-area: preview;
        color: #777;
        font-size: 21px;
        
        .dimmed {
            color: gainsboro;
        }
    }
    
    //.chart,
    .info {
        margin-top: $grid-gap;
        overflow: auto;
    }

    .info {
        grid-area: info;
        -webkit-border-radius: 14px;
        -moz-border-radius: 14px;
        border-radius: 14px;
        overflow: hidden;
        position: relative;
        min-height: 70%;
    }
    
    //.chart {
    //    grid-area: chart;
    //
    //    .vowels {
    //        grid-template-columns: auto auto auto;
    //    }
    //
    //    .consonants {
    //        grid-template-columns: auto auto auto auto auto auto auto;
    //    }
    //
    //    .vowels,
    //    .consonants {
    //        display: grid;
    //        grid-column-gap: $grid-gap;
    //        justify-content: center;
    //        justify-items: center;
    //
    //        > div {
    //            @include transition('ease-in-out', 0.3s);
    //            cursor: pointer;
    //            text-align: center;
    //            width: 70px;
    //            background-color: white;
    //            color: black;
    //            font-size: 28px;
    //            border-radius: 14px;
    //            margin-top: $grid-gap;
    //            box-shadow: 0 0 7px #000;
    //
    //            > div:last-child {
    //                border-top: 1px solid #777;
    //                color: #777;
    //                font-size: 14px;
    //                font-weight: bold;
    //            }
    //
    //            div {
    //                padding: 7px;
    //            }
    //        }
    //
    //        > div:hover {
    //            transform: scale(1.3)
    //        }
    //    }
    //}
}