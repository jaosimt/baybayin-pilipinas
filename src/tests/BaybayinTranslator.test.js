
import BaybayinTranslator from "../components/Content/BaybayinTranslator";

const wrapper = mount(
    <BaybayinTranslator/>
);

it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

it('renders a textarea, a div preview and a div info', () => {
    expect(wrapper.find('textarea').length).toBe(1);
    expect(wrapper.find('.preview').length).toBe(1);
    expect(wrapper.find('.info').length).toBe(1);
});

it('has a working baybayin unicode keyboard with a working translation', () => {
    wrapper.find('textarea').simulate('keyUp', {
        target: { value: 'baybayin' }
    });
    
    expect(wrapper.find('textarea').text()).toBe('ᜊᜌ᜔ᜊᜌᜒᜈ᜔');
    expect(wrapper.find('.preview').text()).toBe('baybay[i/e]n');
});