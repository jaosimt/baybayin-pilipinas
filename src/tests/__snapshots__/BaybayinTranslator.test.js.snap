// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="baybayin-translator"
>
  <textarea
    defaultValue=""
    onKeyUp={[Function]}
  />
  <div
    className="preview"
    dangerouslySetInnerHTML={
      Object {
        "__html": "← ← ← type or paste something there<br/>and see the translation here. ↓↓↓<br/><br/><i style=\\"color: #ff9900\\"><b>Note:</b><br/>This is an assisted and automatic translator as you type.  No need to worry on the convention of typing on a baybayin keyboard.  Just type normally as you would.</i>",
      }
    }
  />
  <div
    className="info"
  >
    <div
      className="keyboard"
    >
      <div
        className="key-tilde red-ish"
        data-down="\`"
        data-up="~"
        onClick={[Function]}
      >
        <div>
          ~
          <span
            className="pink-fg"
          >
            ᜈ᜔ᜌ
          </span>
        </div>
        <div>
          \`
        </div>
      </div>
      <div
        className="key-1 gray-ish"
        data-down="1"
        data-up="!"
        onClick={[Function]}
      >
        <div>
          !
        </div>
        <div>
          1
        </div>
      </div>
      <div
        className="key-2 gray-ish"
        data-down="2"
        data-up="@"
        onClick={[Function]}
      >
        <div>
          @
        </div>
        <div>
          2
        </div>
      </div>
      <div
        className="key-3 gray-ish"
        data-down="3"
        data-up="#"
        onClick={[Function]}
      >
        <div>
          #
        </div>
        <div>
          3
        </div>
      </div>
      <div
        className="key-4 gray-ish"
        data-down="4"
        data-up="₱"
        onClick={[Function]}
      >
        <div>
          $
          <span
            className="red-fg"
          >
            ₱
          </span>
        </div>
        <div>
          4
        </div>
      </div>
      <div
        className="key-5 gray-ish"
        data-down="5"
        data-up="%"
        onClick={[Function]}
      >
        <div>
          %
        </div>
        <div>
          5
        </div>
      </div>
      <div
        className="key-6 gray-ish"
        data-down="6"
        data-up="^"
        onClick={[Function]}
      >
        <div>
          ^
        </div>
        <div>
          6
        </div>
      </div>
      <div
        className="key-7 gray-ish"
        data-down="7"
        data-up="&"
        onClick={[Function]}
      >
        <div>
          &
        </div>
        <div>
          7
        </div>
      </div>
      <div
        className="key-8 gray-ish"
        data-down="8"
        data-up="*"
        onClick={[Function]}
      >
        <div>
          *
        </div>
        <div>
          8
        </div>
      </div>
      <div
        className="key-9 gray-ish"
        data-down="9"
        data-up="("
        onClick={[Function]}
      >
        <div>
          (
        </div>
        <div>
          9
        </div>
      </div>
      <div
        className="key-0 gray-ish"
        data-down="0"
        data-up=")"
        onClick={[Function]}
      >
        <div>
          )
        </div>
        <div>
          0
        </div>
      </div>
      <div
        className="key-dash gray-ish"
        data-down="-"
        data-up="_"
        onClick={[Function]}
      >
        <div>
          _
        </div>
        <div>
          -
        </div>
      </div>
      <div
        className="key-equal blue-ish"
        data-down="="
        data-up="+"
        onClick={[Function]}
      >
        <div>
          +
        </div>
        <div>
          =
          <img
            alt=""
            src="virama.png"
          />
        </div>
      </div>
      <div
        className="key-backspace gray-bg"
        onClick={[Function]}
      >
        <div />
        <div>
          BackSpace
        </div>
      </div>
      <div
        className="key-tab gray-bg"
        data-down="	"
        data-up="	"
        onClick={[Function]}
      >
        <div />
        <div>
          Tab
        </div>
      </div>
      <div
        className="key-q red-ish"
        data-down="q"
        data-up="Q"
        onClick={[Function]}
      >
        <div>
          Q
          <span
            className="pink-fg"
          >
            ᜃ᜔ᜏ
          </span>
        </div>
        <div>
          q
          <span
            className="pink-fg"
          >
            ᜃᜓᜏ
          </span>
        </div>
      </div>
      <div
        className="key-w green-ish"
        data-down="w"
        data-up="W"
        onClick={[Function]}
      >
        <div>
          W
        </div>
        <div>
          w
          <span
            className="pink-fg"
          >
            ᜏ
          </span>
        </div>
      </div>
      <div
        className="key-e blue-ish"
        data-down="e"
        data-up="E"
        onClick={[Function]}
      >
        <div>
          E
          <span
            className="pink-fg"
          >
            ᜁ
          </span>
        </div>
        <div>
          e
          <img
            alt=""
            src="kudlit-e.png"
          />
        </div>
      </div>
      <div
        className="key-r red-ish"
        data-down="r"
        data-up="R"
        onClick={[Function]}
      >
        <div>
          R
        </div>
        <div>
          r
          <span
            className="pink-fg"
          >
            ᜍ
          </span>
        </div>
      </div>
      <div
        className="key-t green-ish"
        data-down="t"
        data-up="T"
        onClick={[Function]}
      >
        <div>
          T
        </div>
        <div>
          t
          <span
            className="pink-fg"
          >
            ᜆ
          </span>
        </div>
      </div>
      <div
        className="key-y green-ish"
        data-down="y"
        data-up="Y"
        onClick={[Function]}
      >
        <div>
          Y
        </div>
        <div>
          y
          <span
            className="pink-fg"
          >
            ᜌ
          </span>
        </div>
      </div>
      <div
        className="key-u blue-ish"
        data-down="u"
        data-up="U"
        onClick={[Function]}
      >
        <div>
          U
          <span
            className="pink-fg"
          >
            ᜂ
          </span>
        </div>
        <div>
          u
          <img
            alt=""
            src="kudlit-u.png"
          />
        </div>
      </div>
      <div
        className="key-i blue-ish"
        data-down="i"
        data-up="I"
        onClick={[Function]}
      >
        <div>
          I
          <span
            className="pink-fg"
          >
            ᜁ
          </span>
        </div>
        <div>
          i
          <img
            alt=""
            src="kudlit-i.png"
          />
        </div>
      </div>
      <div
        className="key-o blue-ish"
        data-down="o"
        data-up="O"
        onClick={[Function]}
      >
        <div>
          O
          <span
            className="pink-fg"
          >
            ᜂ
          </span>
        </div>
        <div>
          o
          <img
            alt=""
            src="kudlit-o.png"
          />
        </div>
      </div>
      <div
        className="key-p green-ish"
        data-down="p"
        data-up="P"
        onClick={[Function]}
      >
        <div>
          P
        </div>
        <div>
          p
          <span
            className="pink-fg"
          >
            ᜉ
          </span>
        </div>
      </div>
      <div
        className="key-leftbracket gray-ish"
        data-down="["
        data-up="{"
        onClick={[Function]}
      >
        <div>
          {
        </div>
        <div>
          [
        </div>
      </div>
      <div
        className="key-rightbracket gray-ish"
        data-down="]"
        data-up="}"
        onClick={[Function]}
      >
        <div>
          }
        </div>
        <div>
          ]
        </div>
      </div>
      <div
        className="key-backslash gray-ish"
        data-down="\\\\"
        data-up="|"
        onClick={[Function]}
      >
        <div>
          |
        </div>
        <div>
          \\
        </div>
      </div>
      <div
        className="key-capslock gray-bg"
        onClick={[Function]}
      >
        <div
          className="capslight"
          style={
            Object {
              "color": "#fff",
            }
          }
        >
          •
        </div>
        <div>
          Capslock
        </div>
      </div>
      <div
        className="key-a blue-ish"
        data-down="a"
        data-up="A"
        onClick={[Function]}
      >
        <div>
          A
          <span
            className="pink-fg"
          >
            ᜀ
          </span>
        </div>
        <div>
          a
        </div>
      </div>
      <div
        className="key-s green-ish"
        data-down="s"
        data-up="S"
        onClick={[Function]}
      >
        <div>
          S
        </div>
        <div>
          s
          <span
            className="pink-fg"
          >
            ᜐ
          </span>
        </div>
      </div>
      <div
        className="key-d green-ish"
        data-down="d"
        data-up="D"
        onClick={[Function]}
      >
        <div>
          D
        </div>
        <div>
          d
          <span
            className="pink-fg"
          >
            ᜇ
          </span>
        </div>
      </div>
      <div
        className="key-f red-ish"
        data-down="f"
        data-up="F"
        onClick={[Function]}
      >
        <div>
          F
        </div>
        <div>
          f
          <span
            className="pink-fg"
          >
            ᜉ
          </span>
        </div>
      </div>
      <div
        className="key-g green-ish"
        data-down="g"
        data-up="G"
        onClick={[Function]}
      >
        <div>
          G
        </div>
        <div>
          g
          <span
            className="pink-fg"
          >
            ᜄ
          </span>
        </div>
      </div>
      <div
        className="key-h green-ish"
        data-down="h"
        data-up="H"
        onClick={[Function]}
      >
        <div>
          H
        </div>
        <div>
          h
          <span
            className="pink-fg"
          >
            ᜑ
          </span>
        </div>
      </div>
      <div
        className="key-j red-ish"
        data-down="j"
        data-up="J"
        onClick={[Function]}
      >
        <div>
          J
          <span
            className="pink-fg"
          >
            ᜇ᜔ᜌ
          </span>
        </div>
        <div>
          j
          <span
            className="pink-fg"
          >
            ᜇᜒᜌ
          </span>
        </div>
      </div>
      <div
        className="key-k green-ish"
        data-down="k"
        data-up="K"
        onClick={[Function]}
      >
        <div>
          K
        </div>
        <div>
          k
          <span
            className="pink-fg"
          >
            ᜃ
          </span>
        </div>
      </div>
      <div
        className="key-l green-ish"
        data-down="l"
        data-up="L"
        onClick={[Function]}
      >
        <div>
          L
        </div>
        <div>
          l
          <span
            className="pink-fg"
          >
            ᜎ
          </span>
        </div>
      </div>
      <div
        className="key-colon gray-ish"
        data-down=";"
        data-up=":"
        onClick={[Function]}
      >
        <div>
          :
        </div>
        <div>
          ;
        </div>
      </div>
      <div
        className="key-quote gray-ish"
        data-down="'"
        data-up="\\""
        onClick={[Function]}
      >
        <div>
          "
        </div>
        <div>
          '
        </div>
      </div>
      <div
        className="key-enter gray-bg"
        data-down="
"
        data-up="
"
        onClick={[Function]}
      >
        <div />
        <div>
          Enter
        </div>
      </div>
      <div
        className="key-leftshift gray-bg"
        onClick={[Function]}
      >
        <div />
        <div>
          Shift
        </div>
      </div>
      <div
        className="key-z red-ish"
        data-down="z"
        data-up="Z"
        onClick={[Function]}
      >
        <div>
          Z
        </div>
        <div>
          z
          <span
            className="pink-fg"
          >
            ᜐ
          </span>
        </div>
      </div>
      <div
        className="key-x red-ish"
        data-down="x"
        data-up="X"
        onClick={[Function]}
      >
        <div>
          X
        </div>
        <div>
          x
          <span
            className="pink-fg"
          >
            ᜃ᜔ᜐ
          </span>
        </div>
      </div>
      <div
        className="key-c red-ish"
        data-down="c"
        data-up="C"
        onClick={[Function]}
      >
        <div>
          C
          <span
            className="pink-fg"
          >
            ᜆ᜔ᜐ
          </span>
        </div>
        <div>
          c
          <span
            className="pink-fg"
          >
            ᜃ
          </span>
        </div>
      </div>
      <div
        className="key-v red-ish"
        data-down="v"
        data-up="V"
        onClick={[Function]}
      >
        <div>
          V
        </div>
        <div>
          v
        </div>
      </div>
      <div
        className="key-b green-ish"
        data-down="b"
        data-up="B"
        onClick={[Function]}
      >
        <div>
          B
        </div>
        <div>
          b
          <span
            className="pink-fg"
          >
            ᜊ
          </span>
        </div>
      </div>
      <div
        className="key-n green-ish"
        data-down="n"
        data-up="N"
        onClick={[Function]}
      >
        <div>
          N
          <span
            className="pink-fg"
          >
            ᜅ
          </span>
        </div>
        <div>
          n
          <span
            className="pink-fg"
          >
            ᜈ
          </span>
        </div>
      </div>
      <div
        className="key-m green-ish"
        data-down="m"
        data-up="M"
        onClick={[Function]}
      >
        <div>
          M
        </div>
        <div>
          m
          <span
            className="pink-fg"
          >
            ᜋ
          </span>
        </div>
      </div>
      <div
        className="key-comma gray-ish"
        data-down=","
        data-up="<"
        onClick={[Function]}
      >
        <div>
          &lt;
        </div>
        <div>
          ,
        </div>
      </div>
      <div
        className="key-period gray-ish"
        data-down="."
        data-up=">"
        onClick={[Function]}
      >
        <div>
          &gt;
        </div>
        <div>
          .
        </div>
      </div>
      <div
        className="key-slash gray-ish"
        data-down="/"
        data-up="?"
        onClick={[Function]}
      >
        <div>
          ?
        </div>
        <div>
          /
        </div>
      </div>
      <div
        className="key-rightshift gray-bg"
        onClick={[Function]}
      >
        <div />
        <div>
          Shift
        </div>
      </div>
      <div
        className="key-leftctrl gray-bg"
      >
        <div />
        <div>
          Control
        </div>
      </div>
      <div
        className="key-empty one gray-bg"
      />
      <div
        className="key-leftalt gray-bg"
      >
        <div />
        <div>
          Alt
        </div>
      </div>
      <div
        className="key-spacebar gray-ish"
        data-down=" "
        data-up=" "
        onClick={[Function]}
      />
      <div
        className="key-rightalt gray-bg"
      >
        <div />
        <div>
          Alt
        </div>
      </div>
      <div
        className="key-empty two gray-bg"
      />
      <div
        className="key-rightctrl gray-bg"
      >
        <div />
        <div>
          Control
        </div>
      </div>
    </div>
  </div>
</div>
`;
